### PR TITLE
Feat: consume per partition

### PR DESCRIPTION
**Pull request by martijnimhoff**
Pull request checklist:

- [x] Include new test cases (either end-to-end or unit tests) with your change.
  > Existing test cases do not include consuming via the standard API. So did not make those for this case.
- [x] Follow our style guides.
- [x] Make sure all tests are still passing and the linter does not report any issues.
  > Current master has quite a few linter errors. I made sure to not add any new errors and I've fixed a few.
- [x] End files with a new line.
- [x] Document the new code in the comments (if it is JavaScript) so the documentation generator can update the reference documentation.
- [x] Avoid platform-dependent code.
- [x] Note: If making modifications to the underlying C++, please use built-in precompiler directives to detect such platform specificities. Use Nan whenever possible to abstract node/v8 version incompatibility.
- [x] Make sure your branch is up to date and rebased.
- [x] Squash extraneous commits unless their history truly adds value to the library.

### CODE REVIEW

**Comment by martijnimhoff** [comment]
@jdbruijn @everhardt Could you review before we submit the PR to Blizzard?

**Comment by jdbruijn** [comment]
Yes, but will be next week.

**Comment by everhardt** [comment]
This implementation allows consuming some partitions individually and leaving the others in the global queue. The current way of doing that has a few weird consequences:
1. once you called `.consume(..)` for a specific partition, it cannot be brought back into the `consumer.consume` of the global queue. I practice I don't think this is an issue and a small function `.forwardToGlobalQueue(partitionId)` can always be added
2. if you don't manually call `.consume(..)` on the global queue you won't get any rebalance events. The alternative of doing this automatically somewhere seems less attractive to me
3. if you don't listen out the rebalance events, some partitions will remain be in the global queue and will thus be part of the callback of the `.consume(..)` call of the global queue, probably even in flowing mode.

File: lib/kafka-consumer.js [hunk 1]
```diff
@@ -361,6 +361,20 @@ KafkaConsumer.prototype.unsubscribe = function() {
...
+ *
+ * Can be useful if the consume performance differs per partition. Consuming
+ * per partition could prevent slow performance on one partition to effect the
+ * consumption of other partitions.
+ *
+ * To select the right partition it is required to set a topic param, because a
+ * consumer can be subscribed to multiple topics.
+ *
+ * @param {number} size - Number of messages to read
+ * @param {number} topic - Name of topic to read
```
Comment (jdbruijn) [inline lib/kafka-consumer.js]: This is a string.

File: src/workers.cc [hunk 2]
```diff
@@ -511,8 +511,7 @@ void ProducerSendOffsetsToTransaction::Execute() {
   Baton b = producer->SendOffsetsToTransaction(
     m_topic_partitions,
     consumer,
-    m_timeout_ms
-  );
+    m_timeout_ms);
```
Comment (jdbruijn) [inline src/workers.cc:514]: Did this change come from the linter (ccplint.py)? If not, this should probably not be changed, doesn't hurt either of course as it's just styling.

File: src/workers.cc [hunk 3]
```diff
@@ -824,9 +991,10 @@ void KafkaConsumerConsumeNum::Execute() {
           if (m_messages.size() > eof_event_count) {
             timeout_ms = 1;
           }
-          
-          // We will only go into this code path when `enable.partition.eof` is set to true
-          // In this case, consumer is also interested in EOF messages, so we return an EOF message
+
+          // We will only go into this code path when `enable.partition.eof` is
+          // set to true. In this case, consumer is also interested in EOF
+          // messages, so we return an EOF message
```
Comment (jdbruijn) [inline src/workers.cc:997]: Did this change come from the linter (ccplint.py)? If not, this should probably not be changed, doesn't hurt either of course as it's just styling.

File: src/kafka-consumer.cc [hunk 4]
```diff
@@ -1122,27 +1123,65 @@ NAN_METHOD(KafkaConsumer::NodeConsume) {
...
-    v8::Local<v8::Number> numMessagesNumber = info[1].As<v8::Number>();
-    Nan::Maybe<uint32_t> numMessagesMaybe = Nan::To<uint32_t>(numMessagesNumber);  // NOLINT
+      KafkaConsumer* consumer = ObjectWrap::Unwrap<KafkaConsumer>(info.This());
 
-    uint32_t numMessages;
-    if (numMessagesMaybe.IsNothing()) {
-      return Nan::ThrowError("Parameter must be a number over 0");
+      v8::Local<v8::Function> cb = info[4].As<v8::Function>();
+      Nan::Callback *callback = new Nan::Callback(cb);
+
```
Comment (jdbruijn) [inline src/kafka-consumer.cc]: Remove new-line to stay in line with the other code, e.g. below at L1181.

File: lib/kafka-consumer.js [hunk 5]
```diff
@@ -361,6 +361,20 @@ KafkaConsumer.prototype.unsubscribe = function() {
 };
 
 /**
+ * Read a number of messages from a specific topic and partition.
+ *
+ * Can be useful if the consume performance differs per partition. Consuming
+ * per partition could prevent slow performance on one partition to effect the
+ * consumption of other partitions.
```
Comment (jdbruijn) [inline lib/kafka-consumer.js]:
```suggestion
 * Can be useful if the consume performance differs per partition. Consuming
 * per partition could prevent slow performance on one partition from affecting
 * the consumption of other partitions.
```

**Comment by everhardt** [comment]
@martijnimhoff What's the status of this one?

File: src/workers.cc [hunk 6]
```diff
@@ -783,6 +782,174 @@ void KafkaConsumerConsumeLoop::HandleErrorCallback() {
   callback->Call(argc, argv);
 }
 
+/**
+ * @brief KafkaConsumer get messages worker.
```
Comment (jdbruijn) [inline src/workers.cc]:
```suggestion
 * @brief KafkaConsumer get messages per partition worker.
```

**Comment by jdbruijn** [comment]
AFAIK waiting for my review, then we can merge it here and start using this in the Teleport code, as well as opening an PR in the upstream. Review in progress BTW

image

**Review by jdbruijn — CHANGES_REQUESTED** [review]

File: src/workers.cc (hunk 2 above)
Comment (martijnimhoff) [reply src/workers.cc:514]: This is indeed a lint fix. Current master has quite a few linter errors. I made sure to not add any new errors and I've fixed a few.

File: src/workers.cc (hunk 3 above)
Comment (martijnimhoff) [reply src/workers.cc:997]: This is indeed a lint fix. Current master has quite a few linter errors. I made sure to not add any new errors and I've fixed a few.

File: src/kafka-consumer.cc (hunk 4 above)
Comment (martijnimhoff) [reply src/kafka-consumer.cc]: Done.

**Comment by everhardt** [comment]
@martijnimhoff how did you address my [third point](https://github.com/withthegrid/node-rdkafka/pull/2#issuecomment-1891073659)?

**Comment by martijnimhoff** [comment]
> if you don't listen out the rebalance events, some partitions will remain be in the global queue and will thus be part of the callback of the .consume(..) call of the global queue, probably even in flowing mode.

It's a strange situation, but perhaps a use-case would exist in which you want to control the consumption per partition of just a few partitions and the remaining partitions could be consumed via the callback. So I would say it's an odd, but  desired behavior. I'm actually now sure if this will function correctly... I'll do a test to see what happens.

**Comment by martijnimhoff** [comment]
@everhardt it indeed results in data loss.

When the consume per partition starts after a timeout of 1s. All the messages in the partition (50) are already consumed by the regular .consume queue. This can be seen in the `data` event and in the callback. 

Perhaps i can see if can create a consumer configuration which disables the forwarding to the general queue for all partitions. Something like this: 

```
var consumer = new KafkaNode.KafkaConsumer({
   ...
  'consume_per_partition': true,
}, {
   ...
});
```

**Pull request merged**